File: C-Plus-Plus/math/decimal_to_hexadecimal.cpp
#include<bits/stdc++.h>
using namespace std;

//CONVERSION OF DECIMAL NUMBER TO HEXADECIMAL NUMBER
void decimal_to_hexadecimal(int n)
{
  int i=0;
  string ans="";
  int last_digit=0;
  char x;
  while(n)
  {
    last_digit=n%16;
    n=n/16;
    /*  Condition checking :
        If digit greater than 9, X will be alphabet
        Else X will be a number  */
    if(last_digit>9)
       x=last_digit+'A'-10;
    else
      x=last_digit+'0';
    ans[i]=x;
    i++;
  }
  //Loop for printing reverse String
  cout<<"\n The Required Hexadecimal Number is::";
  for(int k=i-1;k>=0;k--)
    cout<<ans[k];
}
int main()
{
  int decimal;
  cout<<"\n ENTER THE DECIMAL NUMBER::";
  cin>>decimal;
  decimal_to_hexadecimal(decimal);
  return 0;
}

/* Testcases:

  INPUT-
  ENTER THE DECIMAL NUMBER::463
  OUTPUT-
  The Required Hexadecimal Number is::1CF

  INPUT-
  ENTER THE DECIMAL NUMBER::100
  OUTPUT-
  The Required Hexadecimal Number is::64

  INPUT-
  ENTER THE DECIMAL NUMBER::516042
  OUTPUT-
  The Required Hexadecimal Number is::7DFCA

  Time Complexity- O(n);
  Space Complexity- O(1);

*/

File: Dart/maths/decimal_to_hexadecimal.dart
/*

  Conversion from decimal to Hexadecimal number in dart

*/

import "dart:io";

void decimalToHexadecimal(int decimalNumber) {
  List<int> ans = [];

  int lastDigit = 0, hexCodeLen = 0;

  while (decimalNumber != 0) {
    lastDigit = decimalNumber % 16;
    decimalNumber = (decimalNumber / 16).floor();

    /*
     If lastDigit is greater than 9 then X will be alphabet(hexadecimal),
     else X will be a number(hexadecimal)
     */
    int hexCharCode;
    if (lastDigit > 9)
      hexCharCode = lastDigit + 'A'.codeUnitAt(0) - 10;
    else
      hexCharCode = lastDigit + '0'.codeUnitAt(0);

    ans.add(hexCharCode);
    hexCodeLen++;
  }

  //Loop for printing String in Reverse order
  stdout.write("Hexadecimal Number: ");
  for (int k = hexCodeLen - 1; k >= 0; k--) {
    stdout.write(String.fromCharCode(ans[k]));
  }
}

int main() {
  stdout.write("Enter the Decimal Number: ");

  int decimalNumber = int.parse(stdin.readLineSync()!);
  decimalToHexadecimal(decimalNumber);

  return 0;
}

/* Testcases:

  Test 1
  Enter the Decimal Number: 3452
  Hexadecimal Number: D7C

  Test 2
  Enter the Decimal Number:100
  Hexadecimal Number: 64

  Test 3
  Enter the Decimal Number: 12341234244333 
  Hexadecimal Number: B396AE21AED

  Time Complexity- O(n);
  Space Complexity- O(1);

*/
